### DISCUSSION TITLE
React, Express and Pocketbase

### DISCUSSION THREAD

**Discussion by 4rm1n** [pocketbase/pocketbase#2645] [Q&A] [answered]
Hi,

I'm creating a webapp in react and pocketbase and the React application integrates with Pocketbase thanks to the SDK.
I've already deployed Pocketbase on DigitalOcean and it's accessible via the React app.

For some functions, certain requests have to go through a server side that I've written in node (express). Since the deployment on digital ocean, the express server can no longer communicate with Pocketbase, returning the same error every time:

```
cause: Error: write EPROTO 8684856000:error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1546:SSL alert number 80
  
      at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16) {
    errno: -100,
    code: 'EPROTO',
    syscall: 'write'
}
```

How can I get Express to communicate with Pocketbase?

**Comment by ganigeorgiev** [comment]
I'm not sure what could be the cause of the issue and I don't think it is PocketBase related.

 Since both node and PocketBase are on the same server, I assume that you have a reverse proxy (nginx, cady, apache, etc.). Could you share its configuration? Do you use http or https when connecting to PocketBase? _Google-ing the error indicates maybe that you are trying to connect with https to a http-only server and/or server with older TLS (eg. < 1.3) certificate._

**Comment by 4rm1n** [comment]
the node server is in localhost and pocketbase is all alone in its Droplet on Digital ocean. 

I use https and the following configuration was used:
```
[Unit]
Description = pocketbase

[Service]
Type           = simple
User           = root
Group          = root
LimitNOFILE    = 4096
Restart        = always
RestartSec     = 5s
StandardOutput = append:/root/pb/errors.log
StandardError  = append:/root/pb/errors.log
ExecStart      = /root/pb/pocketbase serve --http="mydomain:80" --https="mydomain:443"

[Install]
WantedBy = multi-user.target
```

**Reply by ganigeorgiev** [reply]
Do you have access to the PocketBase DO server from your localhost?

How do you initialize the JS SDK - `new PocketBase("https://mydomain.com")`?

**Reply by 4rm1n** [reply]
on my React app, Pocketbase is initialized that way and it works perfectly

```
import PocketBase from 'pocketbase';

const pocketbase = new PocketBase('https://mydomain.com');
```
Pocketbase backend is reachable via https://mydomain.com/_  
mydomain.com being replaced my the real domain of course

For requests that have to go through the node server, I don't use the SDK provided, as I want the user's token to follow. 

an example is worth a thousand words :)

Post request between React and Express
```
export async function addParticipation(data){
    const participation = await axios.post('http://localhost:3010/api/participation/', data, {
        headers: {
            'Content-Type': 'application/json',
            'Authorization': pocketbase.authStore.token,
        }
    });
    return participation;
}
```
Post request between Express and Pocketbase
```
app.post('/api/participation/', async (req, res) => {
  try {
    // ...
    // Some Logic
    // ...
    const response = await postParticipation(req);
    return res.status(200).json(response.data);
  } catch (error) {
    res.status(500).json({ error: 'Internal Server Error' });
  }
});
```
and the postParticipation server Side looks like that
```
export async function postParticipation(req) {
    return await axiosInstance.post(`https://mydomain.com/api/collections/participations/records`, req.body, {
        headers: req.headers,
    });
}
```

**Reply by ganigeorgiev** [reply]
> I don't use the SDK provided, as I want the user's token to follow.

You can load the token into the JS SDK AuthStore by doing something like:

```js
pb.authStore.save(token, null);
```

But I'm still not sure what is the cause of the issue in your case. You can try connecting to the live demo https://pocketbase.io and see if the same error will be returned (I've tested locally with Node 18.10 and it is working fine for me).

Another test you can try to rule out PocketBase issue is to send a plain curl request to your endpoint.

In general we use mostly the default Go [`tls.Config` options](https://pkg.go.dev/crypto/tls#Config). The minimum default supported version when acting as server is v1.0, but it supports up to v1.3 when the other side initiate it:

```
// By default, TLS 1.2 is currently used as the minimum when acting as a
// client, and TLS 1.0 when acting as a server. TLS 1.0 is the minimum
// supported by this package, both as a client and as a server.
```

You can verify this by running something like:
```sh
openssl s_client -connect mydomain.com:443 -tls1_3
```

**Reply by 4rm1n** [reply]
>  
> You can load the token into the JS SDK AuthStore by doing something like:
> 
> ```js
> pb.authStore.save(token, null);
> ```

I've tried that and it works! Once deployed, won't it create a problem of concurrency when several people send requests at almost the same time since it's _server-side_?

**Reply by ganigeorgiev** [reply, marked as answer]
No, `pb` in your case can't be global. You have to create a unique PocketBase JS SDK instance per request. For example with Express you can use something like [`res.locals`](https://expressjs.com/en/api.html#res.locals) to store the instance. Option to provide a custom headers (and token) per request basis in the SDK will be available after the planned refactoring soon.

**Reply by 4rm1n** [reply]
Thanks for all the information. Everything works as it should now :)
